### PR TITLE
Script Loading Strategy - ML2 #13 - Add `get_eligible_loading_strategy` method

### LINKED ISSUE #13: Addition of `get_eligible_loading_strategy()` method to surface correct strategy for a given script handle
## Description
With the finalisation of the `wp_get_normalized_script_args()` method implementation #10 , including the the script registration process having been updated (see issues [9](https://github.com/10up/wordpress-develop/issues/9) and [8](https://github.com/10up/wordpress-develop/issues/8)), the `get_eligible_loading_strategy()` method can be introduced to the `WP_Scripts` class/API. The current `WP_Scripts` API resides in source [at this location](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php).

This loading strategy will be determined based on a number of factors, such as whether there are dependents or dependencies, and what their intended strategies are. Checks should be performed against all script dependencies to ensure that they too have loading strategies, before actioning the applicable strategy on said script handle. In essence, this should function is a recursive one that must avoid loops.

While this logic will ultimately first get the *intended* loading strategy, i.e the preferred strategy assigned during script registration, it should call any helper methods required to obtains the least strict loading strategy. 

Strategies are ranked from most to least strict as follows:

- async
- defer
- blocking (default)

Async is considered to be the most strict in the sense that it has the most strict requirements that need to be met in order for the strategy to be applied. Blocking, on the other hand, is considered to be the least strict strategy in the sense that if any of a script's dependencies are blocking, it and all its dependencies become blocking as well, and vice versa.

It should also be noted that WordPress core does not attempt to solve for circular dependencies, and this function is not expected to solve for this either.

## Summary of Loading Strategy Rules
* A script has no dependencies or dependents: `async`
* A script has only blocking or deferred dependencies and only deferred dependents without any non standalone “before” position inline scripts: `defer`
* All other scenarios: `blocking` (although not an actual loading attribute, merely the default behaviour)

## Acceptance Criteria
Create new `get_eligible_loading_strategy()` method within the existing [WP_Scripts](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php) class.
* Method to be private as only utilize during script registration process (or prior to printing) in order to gauge most eligible loading strategy.
* Method to accept one parameter, `$handle` (string), which is the string identifier of an already registered script handle. Defaults to `''`.
* Introduction of business logic into the recently added `get_eligible_loading_strategy()` method within the existing [WP_Scripts](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php) class.
* Method should act against the script handle as passed to its $handle parameter,
* Method should check if passed `$handle` exists as a registered script handle, and return a blocking strategy if not, as that is the default.
* Method to utilize the intended loading strategy passed via the applicable script handles `$args` parameter during the use of either `wp_register_script()` or `wp_enqueue_script()` functions, coupled with necessary business logic that checks the dependency tree and adjust the strategy accordingly.
* Method is to be recursive in nature and check all dependents and dependency scripts of a given handle, but should be written in a way that it never enters into an infinite loop.
* Method is to return the most eligible loading and employ the following business rules:
	*  If the script is eligible for an async strategy, the function will apply that since it has the strictest requirements. 
	*  When the intended loading strategy is set to `defer` and the script is eligible for an `async` loading strategy, it should use a  `defer` loading strategy, since if `async` can be used, `defer `certainly can too.
	*  The second potential eligible strategy of a handle is `defer`. If a script is eligible for a `defer` loading strategy, it can use a `blocking` or `defer` loading strategy, but not an `async` loading strategy. When the intended loading strategy for a given script handle is `async` and the script is eligible for a `defer` loading strategy, it should use the `defer` loading strategy. This could be any scenario where an `async` script is depended on by a `defer` script, making the `async` loading strategy on the dependency infeasible.

### Unit Tests
* Unit tests for the various processes script and script dependency logic currently reside in WordPress core source  [at this location](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/dependencies/scripts.php).
* Applicable tests should be added (or existing tests modified) that assert that scripts eligible for an `async` strategy contain the `async` loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added (or existing tests modified) that assert that scripts eligible for a `defer` strategy contain the `defer` loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added (or existing tests modified) that assert that scripts with a blocking (default) strategy contain no loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added to ensure that if `CONCATENATE_SCRIPTS` is set to `true`, that loading strategy logic be bypassed altogether, allowing the WP_Scripts API to function exactly as it did prior to the loading strategy work, as if no loading strategy API was present.
* Test(s) should be introduced to ensure that - when checking dependencies for eligible strategies based on the dependency tree - that a recursive loop is never entered into. This needs to be done due to he fact that multiple dependencies across various scripts may depend on one another, and when checking for eligible strategies there is a chance that a recursive loop be entered into.
	*  Tests should track the total amount of script handles registered, and the number of dependency strategy eligibility  checks should never, in theory, exceed the total amount of registered script handles. 
	* Tests to DOT have to solve for actual circular decencies of scripts themselves, as core currently does not handle this scenario at all, but they should test to ensure that circular dependencies are not introduced as a result of work done in this issue

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kt-12**
Trac ticket: 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [inline src/wp-includes/class-wp-scripts.php]: Removing default value for `strategy` here.  In `get_eligible_loading_strategy` it becomes hard to know if the **blocking** strategy was from the default logic or it was entered by user(one who wrote the code) during script registration/enqueue. This information is needed both in `async` and `defer` logic.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [inline src/wp-includes/class-wp-scripts.php:776]: Note: This needs to be optimised further.  Better way of doing this is creating a dependents list as a hash table (added as a Dependencies to WP_Script) and updated dependents while each script is getting registered. `get_dependent` will simply fetch dependent from that hash table.

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
	 * @return string|bool Strategy set during script registration. False if none was set.
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
		return (bool) $this->get_data( $handle, 'after' );
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php:814]: Readability suggestion.
```suggestion

		// If there are no dependents remaining to consider, the script can be deferred and the branch ends.
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]: Just to be consistent with other docblock descriptions for `$handle`.

```suggestion
	 * @param string  $handle Name of the script.
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]: Tricky one, but [WP inline comment standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments) only include a single asterisk on the first line. 
```suggestion

		/*
		 * Handle known blocking strategy scenarios.
		 *
		 * Blocking if not a registered handle.
		 * If the script has an 'after' inline dependency, then it can't be eligible for async or defer.
		 */
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]: As a formatting suggestion, for multi-line conditional, put the operator at the beginning of each line rather than at the end. I'm wondering if there is ever a case where a non-registered handle would end up here, given that this is called as a private method? If so, then that should probably be checked before trying to get the intended strategy above and bail early. I'd also suggest checking for the 'blocking' value before checking for an after inline dependency, to avoid unnecessary lookups if our already have the info you need to determine that this should be an empty string.

```suggestion
		if (
			'blocking' === $intended_strategy
			|| $this->has_after_inline_dependency( $handle )
		) {
			return '';
		}
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]: I'd love to see some more test cases set up to confirm that each scenario you're accounting for in this logic tree is working as expected.

**Review by joemcgill — COMMENTED** [review]
Leaving feedback (mostly formatting related).

As a process note, I'm find it a bit difficult to follow the way you have PRs set up to merge into milestone branches before merging to the main feature branch. Given that this is meant to be a continuation from #34, I'd love to see us get that merged into the base feature branch you've set up and then set each of these M2 branches to merge into the main feature branch, rather than waiting for all of it to land "at once" as part of a milestone. That way, we're also getting the benefit of seeing the test coverage increase over time, with each PR and will more likely avoid merge conflicts at the end.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: Removed inline and also introduced back opt-in.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: @joemcgill  Test cases are the main reason I had to set a base milestone branch. The way our milestone tasks are, we normally have to get to the last task to test out all the other prior tasks in the milestone. 

In milestone 2, `get_eligible_strategy`  can only be tested via `do_item` method ( https://github.com/10up/wordpress-develop/pull/38 ) which is a part of different task in same milestone. In milestone 2 all test cases are in PR https://github.com/10up/wordpress-develop/pull/38. 

Just to be certain that everything in a single milestone works as intended before it goes to the base branch, i created a separate milestone base branch.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: Brining back opt-in so now 'blocking' will be the default if value is not set.

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]: Slight wording change here, but I'd suggest we change this comment line to:

```
* Get the strategy assigned during script registration.
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php:796]: Just a thought, do we want to return a mixed value of string|bool when we can stick to just string, and return an empty string if no intended strategy was set? Just a thought, as we can stick to one return type and simply check the result for `empty()`. 

Happy either way, cc @joemcgill for your opinion here as well.

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
			// Recursively check all dependents.
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]: Readability reasons:

```suggestion
	/**
	 * Check if all of a scripts dependents are deferrable which is required to maintain execution order.
	 *
	 * @param string $handle  The script handle.
	 * @param array $checked An array of already visited script handles, used to avoid looping recursion.
	 * @return bool True if all dependents are deferrable, false otherwise.
	 */
	private function all_dependents_are_deferrable( $handle, $checked = array() ) {
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
	 * Get the most eligible loading strategy for a script.
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]: Adding this suggestion based on my previous comment on naming of `$visited` vs `$checked`

```suggestion
			if ( ! $this->all_dependents_are_deferrable( $dependent, $checked ) ) {
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
	 * @param string  $handle The registered handle of the script.
```

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [inline src/wp-includes/class-wp-scripts.php]: @kt-12  Do we need to mention `Dependency will never be set async. So only checking dependent.` here? I am not following what this means, did you mean to state that dependencies will never be deferred so you're only checking dependents?

**Review by 10upsimon — CHANGES_REQUESTED** [review]

**Comment by 10upsimon** [comment]
For visibility sake, as I myself wondered where the unit tests were that handled the various loading strategies, they've been added to the PR for #14 and can be seen at https://github.com/10up/wordpress-develop/pull/38/files#diff-09a2e7959959d6a08ef07b1849e0178b853f7ef86f7885bbf9be2ed261145b24

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: @10upsimon  So, per the rules, defer can only have 'defer' or 'blocking' dependencies; it can never be 'async'.  ( point 1 )

A decision that a script is to be set to `async`, `defer`, or nothing is made by `get_eligible_loading_strategy`.

But, if you check the rules for `async` in this function, a script with dependent will never be set `async`.  ( point 2 ). 

From points 1 and 2, I can say if the current script has some dependency, the dependency script will never have `async` strategy (as it has a dependent).  So we can avoid a recursion check to find strategy in dependencies.  

I hope I still don't sound confusing.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [inline src/wp-includes/class-wp-scripts.php]: @10upsimon, here we can check for `async` too. 

`get_eligible_loading_strategy` converts any `async` with dependencies to `defer` ( if in this recursion it only has deferred `dependent`

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php:796]: I am okay with both. I, too leave this to @joemcgill

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]:
```suggestion
	 * @return string|bool Strategy in script registration, False if not strategy is mentioned.
```

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [reply src/wp-includes/class-wp-scripts.php:796]: It's a good question. In this case, I think returning `false` to indicate that no strategy was set is fine here and communicated intent a bit more than an empty string.

File: tests/phpunit/tests/dependencies/scripts.php
Comment (joemcgill) [inline tests/phpunit/tests/dependencies/scripts.php]: Good idea! We can go ahead and set these to `@ticket 12009`.

File: src/wp-includes/class-wp-scripts.php
Comment (joemcgill) [inline src/wp-includes/class-wp-scripts.php]: I'm still confused about why script registration is being checked here after the intended strategy instead of before? Can you explain the use case for me?

I'd also suggest splitting this back to multiple lines for readability, just make sure the operators are at the beginning of each line, like the previous suggestion.

**Review by joemcgill — COMMENTED** [review]
The updates are looking good. I left a couple new questions and will do a final review once we have the tests from #14 merged back here so we can ensure everything is covered.

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [reply src/wp-includes/class-wp-scripts.php:796]: Happy with that then @joemcgill @kt-12 , marking this resolved.

**Comment by 10upsimon** [comment]
> 

@joemcgill  in case you missed it, noting that the PR at https://github.com/10up/wordpress-develop/pull/34 was merged into the same base branch this is going to 👍

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: You are correct. script registration should be above getting intended script. I have made the change. 

Now that the condition have become much smaller I have kept it on the same line.

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [reply src/wp-includes/class-wp-scripts.php]: @10upsimon  I have removed the extra text here.

File: tests/phpunit/tests/dependencies/scripts.php
Comment (kt-12) [reply tests/phpunit/tests/dependencies/scripts.php]: Thank you. I have updated all the ticked id in #38

File: src/wp-includes/class-wp-scripts.php
Comment (kt-12) [inline src/wp-includes/class-wp-scripts.php:822]: @10upsimon  I have added for check for `async` here. Since this is a complex case i have created a separate test case for this `test_defer_with_async_dependent`

**Review by joemcgill — APPROVED** [review]
I think this is looking good now 👍🏻

**Review by 10upsimon — APPROVED** [review]

File: src/wp-includes/class-wp-scripts.php
Comment (10upsimon) [reply src/wp-includes/class-wp-scripts.php:822]: Thanks @kt-12 this looks much better.

**Pull request merged**
